### PR TITLE
`Sort` `Birthday` feature, no tests, transfer code ownership

### PR DESCRIPTION
**Pull request by UdhayaShan1**
# Description:
This pull request implements the "Sort Birthday" feature, which allows users to sort the list of clients by their birthdays. The implementation consists of several commits, each addressing specific aspects of the feature:

## Create SortCommand: Introduces the SortCommand class responsible for executing the sorting operation.
Add Getters for Day, Month, and Year for Birthday: Adds getter methods for accessing the day, month, and year components of a birthday.

Add Getter for Copy of Internal Observable List: Implements a getter method to obtain a copy of the internal observable list of persons in the Realodex.

Add Case for Sort Command in RealodexParser: Updates the RealodexParser to handle the sorting command.

Add Some Helper Getters First, Might Remove: Includes additional helper methods (may be subject to removal/refactoring).

Add Number of Days from Today to Birthday Function: Implements a method to calculate the number of days from the current date to a person's birthday.

Add Update List, Not Able to Unsort Though: Adds functionality to update the sorted list of persons in the Realodex.

Add toString() for Number of Days: Implements a toString() method for displaying the number of days until a person's birthday.

Fix Sorting Bugs, Fix Checkstyles: Addresses sorting issues and ensures compliance with checkstyle rules.

## Changes:

Added the SortCommand class to facilitate sorting by birthday.
Implemented methods to calculate the number of days until a person's birthday.
Updated the parser to handle sorting commands.
Introduced getter methods for accessing birthday components.
Fixed sorting bugs and ensured code compliance with checkstyle rules.
Transferred code ownership as its not really my domain

## Testing:

Manual testing performed to verify sorting functionality.
Test cases to be added in subsequent commits.

### CODE REVIEW

File: src/main/java/seedu/realodex/logic/commands/SortCommand.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * Returns a string representation of this command.
+     *
+     * @return a string representation of this command
+     */
+    @Override
+    public String toString() {
+        return new ToStringBuilder(this)
+                .toString();
```
Comment (appleraincoat) [inline src/main/java/seedu/realodex/logic/commands/SortCommand.java:52]: Don't think there is a point in this method

File: src/main/java/seedu/realodex/model/person/Birthday.java [hunk 2]
```diff
@@ -121,10 +140,64 @@ public String toString() {
      * Format state as text for representation.
      */
     public String toStringWithRepresentation() {
-        if (birthday.isPresent()) {
-            return "Birthday: " + DATE_FORMAT.format(birthday.get());
+        return birthday.map(date -> "Birthday: " + DATE_FORMAT.format(date)).orElse("No specified Birthday.");
+    }
+
+    public Date getDate() {
+        return birthday.orElse(null);
```
Comment (appleraincoat) [inline src/main/java/seedu/realodex/model/person/Birthday.java:147]: No usages of this method, can i remove this?

File: src/main/java/seedu/realodex/model/person/Birthday.java [hunk 3]
```diff
@@ -55,6 +57,23 @@ public Birthday() {
         }
         this.birthday = birthdayDate;
     }
+    /**
+     * Returns the optional birthday date.
+     *
+     * @return The optional birthday date.
+     */
+    public Optional<Date> getBirthday() {
+        return birthday;
```
Comment (appleraincoat) [inline src/main/java/seedu/realodex/model/person/Birthday.java:66]: Tests created.

File: src/main/java/seedu/realodex/model/person/Birthday.java [hunk 4]
```diff
@@ -121,10 +140,64 @@ public String toString() {
...
+    public Date getDate() {
+        return birthday.orElse(null);
+    }
+
+    /**
+     * Returns the number of days from the current system date to the birthday.
+     * If the birthday has already passed this year, it returns the number of days
+     * from the current date of next year to the birthday.
+     */
+    public Long getDaysUntilBirthday() {
```
Comment (appleraincoat) [inline src/main/java/seedu/realodex/model/person/Birthday.java:155]: 👍

File: src/main/java/seedu/realodex/model/person/Birthday.java [hunk 5]
```diff
@@ -121,10 +140,64 @@ public String toString() {
...
+    }
+
+
+
+
+    public String getDaysUntilBirthdayWithRepresentation() {
+        if (this.birthday.isPresent()) {
+            return getDaysUntilBirthday() + " More Days Till Their Birthday!";
         }
+        return "No days till an unspecified birthday!";
```
Comment (appleraincoat) [inline src/main/java/seedu/realodex/model/person/Birthday.java:189]:
```suggestion
        return "Birthday is unspecified!";
```

File: src/main/java/seedu/realodex/model/person/Birthday.java [hunk 6]
```diff
@@ -121,10 +140,64 @@ public String toString() {
...
+
+    private Calendar returnInstanceOfCalendar(Date date) {
+        Calendar toReturn = Calendar.getInstance();
+        toReturn.setTime(date);
+        toReturn.set(Calendar.HOUR_OF_DAY, 0);
+        toReturn.set(Calendar.MINUTE, 0);
+        toReturn.set(Calendar.SECOND, 0);
+        toReturn.set(Calendar.MILLISECOND, 0);
+        return toReturn;
+    }
```
Comment (appleraincoat) [inline src/main/java/seedu/realodex/model/person/Birthday.java:200]: Good abstraction.

File: src/main/java/seedu/realodex/model/person/UniquePersonList.java [hunk 7]
```diff
@@ -104,6 +104,12 @@ public ObservableList<Person> asUnmodifiableObservableList() {
         return internalUnmodifiableList;
     }
 
+    public ObservableList<Person> getInternalListCopy() {
+        ObservableList<Person> copyList = FXCollections.observableArrayList();
+        copyList.addAll(internalList);
+        return copyList;
+    }
+
```
Comment (appleraincoat) [inline src/main/java/seedu/realodex/model/person/UniquePersonList.java:112]: 👍

File: src/main/java/seedu/realodex/model/person/sorting/BirthdayComparator.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+     * If one or both persons have unspecified birthdays, they are sorted accordingly.
+     * If both persons have specified birthdays, they are sorted based on the number of days until their next birthday.
+     *
+     * @param o1 the first person to be compared
+     * @param o2 the second person to be compared
+     * @return a negative integer, zero, or a positive integer as the first person
+     *         is less than, equal to, or greater than the second person, respectively
+     */
+    @Override
+    public int compare(Person o1, Person o2) {
```
Comment (appleraincoat) [inline src/main/java/seedu/realodex/model/person/sorting/BirthdayComparator.java:23]: tests added

**Review by appleraincoat — COMMENTED** [review]
Changes are reflected in my PR for this. Thanks for the effort!

**Comment by UdhayaShan1** [comment]
Thanks @appleraincoat for adding tests, PR will be superseded by his

**Pull request closed without merging**